src/PostgREST/Auth/Cache.hs

```diff
@@ -0,0 +1,124 @@
+
+import PostgREST.AppState    (AppState, getJwtCacheState, getObserver)
+import PostgREST.Auth.Types  (AuthResult (..), JwtCacheState (..))
+import PostgREST.Error       (Error (..))
+import PostgREST.Internal    (recursiveSizeNF)
+import PostgREST.Observation (Observation (..))
+
+import Protolude
+
+-- | Used to retrieve and insert JWT to JWT Cache
+getJWTFromCache :: AppState -> ByteString -> Int -> IO (Either Error AuthResult) -> UTCTime -> IO (Either Error AuthResult)
+getJWTFromCache appState token maxLifetime parseJwt utc = do
+
+  checkCache <- C.lookup jwtCache token
+  authResult <- maybe parseJwt (pure . Right) checkCache
+
+  -- if token not found, add to cache and increment cache size metric
+  case (authResult,checkCache) of
+    -- From comment:
+    -- https://github.com/PostgREST/postgrest/pull/3801#discussion_r1857987914
+    --
+    -- We purge expired cache entries on a cache miss
+    -- The reasoning is that:
+    --
+    -- 1. We expect it to be rare (otherwise there is no point of the cache)
+    -- 2. It makes sure the cache is not growing (as inserting new entries
+    --    does garbage collection)
+    -- 3. Since this is time expiration based cache there is no real risk of
+    --    starvation - sooner or later we are going to have a cache miss.
+
+    (Right res, Nothing) -> do -- cache miss
+
+      let timeSpec = getTimeSpec res maxLifetime utc
+
+      -- purge expired cache entries (VERY INEFFICIENT)
+      C.purgeExpired jwtCache
+
+      -- insert new cache entry
+      C.insert' jwtCache (Just timeSpec) token res
+
+      -- calculate cache size on a new thread
+      updateJwtCacheSizeMetric appState
+
+    _                    -> pure ()
+
+  return authResult
+    where
+      jwtCache = cache $ getJwtCacheState appState
+
+-- | Used to extract JWT exp claim and add to JWT Cache
+getTimeSpec :: AuthResult -> Int -> UTCTime -> TimeSpec
+getTimeSpec res maxLifetime utc = do
+  let expireJSON = KM.lookup "exp" (authClaims res)
+      utcToSecs = floor . nominalDiffTimeToSeconds . utcTimeToPOSIXSeconds
+      sciToInt = fromMaybe 0 . Sci.toBoundedInteger
+  case expireJSON of
+    Just (JSON.Number seconds) -> TimeSpec (sciToInt seconds - utcToSecs utc) 0
+    _                          -> TimeSpec (fromIntegral maxLifetime :: Int64) 0
+
+-- | Update JwtCacheSize Metric
+--
+-- Running the cache size calculation on a new thread.
+-- To make this thread safe, a semaphore is used. It
+-- allows only 1 thread at a time to run the calculation blocking
+-- new threads until signaled
+updateJwtCacheSizeMetric :: AppState -> IO ()
+updateJwtCacheSizeMetric appState = do
+  _ <- forkIO $ bracket_ (waitQSem sem) (signalQSem sem) $ do
+    size <- calcCacheSizeInBytes jwtCache
+    observer $ JWTCache size
+  return ()
+      where
+        sem = (semaphore . getJwtCacheState) appState
+        jwtCache = (cache . getJwtCacheState) appState
+        observer = getObserver appState
```

Comment:
Here is the implementation with a background thread, It isn't quite mergeable yet (still needs some refactoring and documenting internals). This PR is getting bigger, perhaps maybe we introduce a `Cache.hs` in a separate PR?

Reply:
> This PR is getting bigger, perhaps maybe we introduce a Cache.hs in a separate PR

Sounds good :+1:.